src/kernel/sporadic.c

```diff
@@ -248,94 +252,60 @@ static bool_t refill_head_overlapping(sched_context_t *sc)
+            .rAmount = usage,
+            .rTime = refill_head(sc)->rTime + sc->scPeriod,
+        };
+
+        refill_head(sc)->rAmount -= usage;
+        /* We need to keep the head refill no more than a period before
+         * the start of the tail refill. This ensures that new refills
+         * are never added before the tail refill (breaking the ordered
+         * invariant). This code actually keeps the head refill no more
+         * than a period before the end of the tail refill (which is
+         * stronger than necessary) but is what is used in the current
+         * proofs. In combination with the merging behaviour of
+         * schedule_used, the following will still ensure that
+         * incremental charging of a refill across preemptions only
+         * produces a single new refill one period in the future. */
+        refill_head(sc)->rTime += usage;
+        schedule_used(sc, used);
     }
 
-    ensure_sufficient_head(sc);
-
-    REFILL_SANITY_END(sc);
-}
-
-void refill_split_check(ticks_t usage)
-{
-    sched_context_t *sc = NODE_STATE(ksCurSC);
-    /* invalid to call this on a NULL sc */
-    assert(sc != NULL);
-    /* something is seriously wrong if this is called and no
-     * time has been used */
-    assert(usage > 0);
-    assert(usage <= refill_head(sc)->rAmount);
-    assert(sc->scPeriod > 0);
-
-    REFILL_SANITY_START(sc);
+    /* Ensure the head refill has the minimum budget */
+    while (refill_head(sc)->rAmount < MIN_BUDGET) {
+        refill_t head = refill_pop_head(sc);
+        refill_head(sc)->rAmount += head.rAmount;
+        /* Delay head to ensure the subsequent refill doesn't end any
+         * later (rather than simply combining refills). */
+        refill_head(sc)->rTime -= head.rAmount;
+    }
 
-    /* first deal with the remaining budget of the current replenishment */
-    ticks_t remnant = refill_head(sc)->rAmount - usage;
-
-    /* set up a new replenishment structure */
-    refill_t new = (refill_t) {
-        .rAmount = usage, .rTime = refill_head(sc)->rTime + sc->scPeriod
-    };
-
-    if (refill_size(sc) == sc->scRefillMax || remnant < MIN_BUDGET) {
-        /* merge remnant with next replenishment - either it's too small
-         * or we're out of space */
-        if (refill_single(sc)) {
-            /* update inplace */
-            new.rAmount += remnant;
-            *refill_head(sc) = new;
-        } else {
-            refill_pop_head(sc);
-            refill_head(sc)->rAmount += remnant;
-            schedule_used(sc, new);
-            ensure_sufficient_head(sc);
-        }
-        assert(refill_ordered(sc));
-    } else {
-        /* leave remnant as reduced replenishment */
-        assert(remnant >= MIN_BUDGET);
-        /* split the head refill  */
-        refill_head(sc)->rAmount = remnant;
-        schedule_used(sc, new);
+    /* Ensure that the refills are now not disjoint */
+    while (unlikely(refill_head_overlapping(sc))) {
```

Comment:
When would this condition become true? From what I can tell, any other operation leaves the queue without any overlapping refills and any changes to the queue earlier in this function also leaves the queue without overlapping entries.

Reply:
For binary verification, it would be very helpful if we could avoid this loop. :-)

Reply:
I also can't see how this loop would ever be executed. I can't see any code that ever creates overlapping refills.

Reply:
(at least not in this PR).

Reply:
I'm not sure as to why this ended up back in here. I also don't think there's any way for overlapping refills to occur.